src/materialized/src/bin/materialized/tracing.rs

```diff
@@ -117,3 +118,72 @@ where
         }
     }
 }
+
+/// A tracing `Layer` that allows hooking into the reporting/filtering chain for log messages,
+/// incrementing a counter for the severity of messages reported.
+pub struct MetricsRecorderLayer<S> {
+    counter: IntCounterVec,
+    _inner: PhantomData<S>,
+}
+
+impl<S> MetricsRecorderLayer<S> {
+    /// Construct a metrics-recording layer.
+    pub fn new(counter: IntCounterVec) -> Self {
+        Self {
+            counter,
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S> Layer<S> for MetricsRecorderLayer<S>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_event(&self, ev: &Event<'_>, _ctx: Context<'_, S>) {
+        let metadata = ev.metadata();
+        self.counter
+            .with_label_values(&[&metadata.level().to_string()])
+            .inc();
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashMap;
+
+    use super::MetricsRecorderLayer;
+    use log::{error, info, warn};
+    use prometheus::{IntCounterVec, Opts, Registry};
+    use tracing_subscriber::prelude::*;
```

Comment:
nit: we try to avoid glob imports even when upstream packages try to sell you on a `prelude` module, just because it makes it hard to trace where structs come from unless you're in an IDE.

Reply:
Fixed - thanks (and agreed, preludes are confusing even if you do have IDE hints) (: